File: src/bigclaw/roadmap.py
from dataclasses import dataclass, field
from typing import Dict, List


@dataclass(frozen=True)
class EpicMilestone:
    epic_id: str
    title: str
    phase: str
    owner: str
    milestone: str


@dataclass
class ExecutionPackRoadmap:
    name: str
    epics: List[EpicMilestone] = field(default_factory=list)

    def phase_map(self) -> Dict[str, List[EpicMilestone]]:
        phases: Dict[str, List[EpicMilestone]] = {}
        for epic in self.epics:
            phases.setdefault(epic.phase, []).append(epic)
        return phases

    def epic_map(self) -> Dict[str, EpicMilestone]:
        return {epic.epic_id: epic for epic in self.epics}

    def validate_unique_owners(self) -> None:
        seen: Dict[str, str] = {}
        for epic in self.epics:
            owner = epic.owner.strip().lower()
            if owner in seen:
                raise ValueError(
                    f"Owner '{epic.owner}' is assigned to both {seen[owner]} and {epic.epic_id}"
                )
            seen[owner] = epic.epic_id


def build_execution_pack_roadmap() -> ExecutionPackRoadmap:
    roadmap = ExecutionPackRoadmap(
        name="BigClaw v5.0 Parallel Distributed Platform Follow-up",
        epics=[
            EpicMilestone(
                epic_id="BIG-EPIC-21",
                title="并行分布式执行平台 closeout baseline",
                phase="Closed Baseline",
                owner="distributed-platform",
                milestone="M0 Closeout evidence complete through review and navigation refresh",
            ),
            EpicMilestone(
                epic_id="BIG-PAR-FU-1",
                title="Replicated event durability rollout",
                phase="Phase 1",
                owner="event-durability",
                milestone="M1 Promote the OPE-222 durability contract into a provider-backed rollout path",
            ),
            EpicMilestone(
                epic_id="BIG-PAR-FU-2",
                title="Remaining hardening gap closure",
                phase="Phase 2",
                owner="coordination-hardening",
                milestone="M2 Close leader election, takeover, and operator-facing hardening gaps",
            ),
            EpicMilestone(
                epic_id="BIG-PAR-FU-3",
                title="External-store and multi-node validation",
                phase="Phase 3",
                owner="platform-validation",
                milestone="M3 Prove the control plane beyond SQLite with higher-scale shared backends",
            ),
            EpicMilestone(
                epic_id="BIG-PAR-FU-4",
                title="Production capacity certification",
                phase="Phase 4",
                owner="release-readiness",
                milestone="M4 Convert local soak evidence into production-grade certification gates",
            ),
        ],
    )
    roadmap.validate_unique_owners()
    return roadmap

File: bigclaw-go/internal/product/roadmap.go
package product

import (
	"fmt"
	"strings"
)

type EpicMilestone struct {
	EpicID    string `json:"epic_id"`
	Title     string `json:"title"`
	Phase     string `json:"phase"`
	Owner     string `json:"owner"`
	Milestone string `json:"milestone"`
}

type ExecutionPackRoadmap struct {
	Name  string          `json:"name"`
	Epics []EpicMilestone `json:"epics"`
}

func (r ExecutionPackRoadmap) PhaseMap() map[string][]EpicMilestone {
	phases := map[string][]EpicMilestone{}
	for _, epic := range r.Epics {
		phases[epic.Phase] = append(phases[epic.Phase], epic)
	}
	return phases
}

func (r ExecutionPackRoadmap) EpicMap() map[string]EpicMilestone {
	epics := make(map[string]EpicMilestone, len(r.Epics))
	for _, epic := range r.Epics {
		epics[epic.EpicID] = epic
	}
	return epics
}

func (r ExecutionPackRoadmap) ValidateUniqueOwners() error {
	seen := map[string]string{}
	for _, epic := range r.Epics {
		owner := strings.ToLower(strings.TrimSpace(epic.Owner))
		if previous, ok := seen[owner]; ok {
			return fmt.Errorf("Owner '%s' is assigned to both %s and %s", epic.Owner, previous, epic.EpicID)
		}
		seen[owner] = epic.EpicID
	}
	return nil
}

func BuildExecutionPackRoadmap() (ExecutionPackRoadmap, error) {
	roadmap := ExecutionPackRoadmap{
		Name: "BigClaw v4.0 Execution Pack",
		Epics: []EpicMilestone{
			{
				EpicID:    "BIG-EPIC-8",
				Title:     "研发自治执行平台增强",
				Phase:     "Phase 1",
				Owner:     "engineering-platform",
				Milestone: "M1 Foundation uplift",
			},
			{
				EpicID:    "BIG-EPIC-9",
				Title:     "工程运营系统",
				Phase:     "Phase 2",
				Owner:     "engineering-operations",
				Milestone: "M2 Operations control plane",
			},
			{
				EpicID:    "BIG-EPIC-10",
				Title:     "跨部门 Agent Orchestration",
				Phase:     "Phase 3",
				Owner:     "orchestration-office",
				Milestone: "M3 Cross-team orchestration",
			},
			{
				EpicID:    "BIG-EPIC-11",
				Title:     "产品化 UI 与控制台",
				Phase:     "Phase 4",
				Owner:     "product-experience",
				Milestone: "M4 Productized console",
			},
			{
				EpicID:    "BIG-EPIC-12",
				Title:     "计费、套餐与商业化控制",
				Phase:     "Phase 5",
				Owner:     "commercialization",
				Milestone: "M5 Billing and packaging",
			},
		},
	}
	if err := roadmap.ValidateUniqueOwners(); err != nil {
		return ExecutionPackRoadmap{}, err
	}
	return roadmap, nil
}
